File: Backend-node/src/controllers/albumController.js
const Album = require("../models/albums");

// Crear un nuevo album
exports.createAlbum = async (req, res) => {
    try {
        const { album_name, id_user } = req.body;

        if (!album_name || !id_user) {
            return res.status(400).json({ message: 'Please fill all fields', status: false });
        }

        const albumExists = await Album.findOne({ where: { album_name } });

        if (albumExists) {
            return res.status(400).json({ message: 'Album already exists', status: false });
        }

        const album = await Album.create({
            album_name,
            id_user
        });

        return res.status(201).json({ message: 'Album created successfully', album, status: true });
    } catch (error) {
        console.error("Error in createAlbum:", error);
        res.status(500).json({ message: 'Server error', error, status: false });
    }
};


exports.getAlbumByUserId = async (req, res) => {
    try {

        const { id_user } = req.params;
        if (!id_user) {
            return res.status(400).json({ message: 'User ID is required', status: false });
        }

        const albums = await Album.findAll({ where: { id_user } });

        if (!albums) {
            return res.status(404).json({ message: 'Album not found', status: false });
        }

        return res.status(200).json({ albums, status: true });
    } catch (error) {
        console.error("Error in get albums by user id:", error);
        return res.status(500).json({ message: 'Server error', error, status: false });
    }
}

// Actualizar un album
exports.updateAlbum = async (req, res) => {
    try {
        const { id_album } = req.params;
        const { album_name } = req.body;

        if (!id_album) {
            return res.status(400).json({ message: 'Album ID is required', status: false });
        }

        const album = await Album.findByPk(id_album);
        
        if (!album) {
            return res.status(404).json({ message: 'Album not found', status: false });
        }
        const albumExists = await Album.findOne({ where: { album_name } });
        if (album.album_name === "Fotos de perfil"){
            return res.status(400).json({ message: 'Album name cannot be changed', status: false });
        }
        if (albumExists) {
            return res.status(400).json({ message: 'Album already exists', status: false });
        }
        
        await Album.update({ album_name }, { where: { id_album } });

        return res.status(200).json({ message: 'Album updated successfully', status: true });
    } catch (error) {
        console.error("Error in updateAlbum:", error);
        return res.status(500).json({ message: 'Server error', error, status: false });
    }
}

// Eliminar un album
exports.deleteAlbum = async (req, res) => {
    try {
        const { id_album } = req.params;

        if (!id_album) {
            return res.status(400).json({ message: 'Album ID is required', status: false });
        }

        const album = await Album.findByPk(id_album);

        if (!album) {
            return res.status(404).json({ message: 'Album not found', status: false });
        }
        if (album.album_name === "Fotos de perfil"){
            return res.status(400).json({ message: 'Album name cannot be deleted', status: false });
        }
        await Album.destroy({ where: { id_album } });

        return res.status(200).json({ message: 'Album deleted successfully', status: true });
    } catch (error) {
        console.error("Error in deleteAlbum:", error);
        return res.status(500).json({ message: 'Server error', error, status: false });
    }
}

File: Backend-python/src/controllers/albumController.py
from flask import request, jsonify
from models.albums import Album
from config.db import get_db

def createAlbum():
    db = next(get_db())
    try:
        # Extraer los datos del cuerpo de la solicitud
        data = request.get_json()
        album_name = data.get('album_name')
        id_user = data.get('id_user')

        # Validar que los campos necesarios están presentes
        if not album_name or not id_user:
            return jsonify({"message": "Please fill all fields", "status": False}), 400

        # Crear el nuevo álbum
        new_album = Album(album_name=album_name, id_user=id_user)
        db.add(new_album)
        db.commit()

        # Retornar la respuesta exitosa
        return jsonify({"message": "Album created successfully", "album": new_album.to_dict(), "status": True}), 201

    except Exception as error:
        db.rollback()  # Revertir la transacción en caso de error
        print(f"Error in createAlbum: {error}")
        return jsonify({"message": "Server error", "error": str(error), "status": False}), 500

def getAlbumByUserId(id_user):
    db = next(get_db())
    try:
        # Validar si el ID del usuario fue proporcionado
        if not id_user:
            return jsonify({"message": "User ID is required", "status": False}), 400

        # Buscar todos los álbumes del usuario especificado
        albums = db.query(Album).filter_by(id_user=id_user).all()

        # Verificar si se encontraron álbumes
        if not albums:
            return jsonify({"message": "Albums not found", "status": False}), 404

        # Convertir los álbumes a diccionarios antes de devolverlos
        albums_list = [album.to_dict() for album in albums]

        return jsonify({"albums": albums_list, "status": True}), 200

    except Exception as error:
        print(f"Error in get albums by user id: {error}")
        return jsonify({"message": "Server error", "error": str(error), "status": False}), 500

def updateAlbum(id_album):
    db = next(get_db())
    try:
        # Obtener los datos de la solicitud
        data = request.json
        album_name = data.get('album_name')

        # Validar si se proporcionó el ID del álbum
        if not id_album:
            return jsonify({"message": "Album ID is required", "status": False}), 400

        # Buscar el álbum por ID
        album = db.query(Album).filter_by(id_album=id_album).first()

        # Verificar si el álbum existe
        if not album:
            return jsonify({"message": "Album not found", "status": False}), 404

        # Comprobar si el nombre del álbum es "Fotos de perfil"
        if album.album_name == "Fotos de perfil":
            return jsonify({"message": "Album name cannot be changed", "status": False}), 400

        # Verificar si ya existe un álbum con el nuevo nombre
        album_exists = db.query(Album).filter_by(album_name=album_name).first()
        if album_exists:
            return jsonify({"message": "Album already exists", "status": False}), 400

        # Actualizar el nombre del álbum
        album.album_name = album_name
        db.commit()

        return jsonify({"message": "Album updated successfully", "status": True}), 200

    except Exception as error:
        print(f"Error in updateAlbum: {error}")
        return jsonify({"message": "Server error", "error": str(error), "status": False}), 500

def deleteAlbum(id_album):
    db = next(get_db())
    try:
        # Validar si se proporcionó el ID del álbum
        if not id_album:
            return jsonify({"message": "Album ID is required", "status": False}), 400

        # Buscar el álbum por su ID
        album = db.query(Album).filter_by(id_album=id_album).first()

        # Verificar si el álbum existe
        if not album:
            return jsonify({"message": "Album not found", "status": False}), 404

        # Verificar si el nombre del álbum es "Fotos de perfil"
        if album.album_name == "Fotos de perfil":
            return jsonify({"message": "Album name cannot be deleted", "status": False}), 400

        # Eliminar el álbum
        db.delete(album)
        db.commit()

        return jsonify({"message": "Album deleted successfully", "status": True}), 200

    except Exception as error:
        print(f"Error in deleteAlbum: {error}")
        return jsonify({"message": "Server error", "error": str(error), "status": False}), 500
